pkg/pool-weighted/contracts/managed/CircuitBreakerStorageLib.sol

```diff
@@ -0,0 +1,322 @@
+     *
+     * As described in the general comments above, the weight complement calculation attempts to isolate changes
+     * in the balance due to arbitrageurs responding to external prices, from internal price changes caused by an
+     * ongoing weight update, or changes to the pool composition. There is a non-linear relationship between "spot"
+     * price changes and BPT price changes. This calculation transforms one into the other.
+     *
+     * @param circuitBreakerState - The bytes32 state of the token of interest.
+     * @param weightComplement - The complement of this token's weight, generally given by (1 - weight).
+     * @return - lower and upper BPT price bounds, which can be directly compared against the current BPT price.
+     */
+    function getCurrentCircuitBreakerBounds(bytes32 circuitBreakerState, uint256 weightComplement)
+        internal
+        pure
+        returns (uint256, uint256)
+    {
+        // Retrieve the weight complement passed in and bptPrice computed when the circuit breaker was set.
+        uint256 bptPrice = circuitBreakerState.decodeUint(_BPT_PRICE_OFFSET, _BPT_PRICE_WIDTH);
+        uint256 initialWeightComplement = circuitBreakerState.decodeUint(
+            _WEIGHT_COMPLEMENT_OFFSET,
+            _WEIGHT_COMPLEMENT_WIDTH
+        );
+
+        uint256 lowerBoundRatio;
+        uint256 upperBoundRatio;
+
+        if (initialWeightComplement == weightComplement) {
+            // If the weight complement hasn't changed since the circuit breaker was set, we can use the precomputed
+            // boundary ratios.
+            lowerBoundRatio = circuitBreakerState.decodeUint(_LOWER_BOUND_RATIO_OFFSET, _BOUND_RATIO_WIDTH).decompress(
+                _BOUND_RATIO_WIDTH,
+                _MAX_BOUND_PERCENTAGE
+            );
+            upperBoundRatio = circuitBreakerState.decodeUint(_UPPER_BOUND_RATIO_OFFSET, _BOUND_RATIO_WIDTH).decompress(
+                _BOUND_RATIO_WIDTH,
+                _MAX_BOUND_PERCENTAGE
+            );
+        } else {
+            // Something has changed - either the weight of the token, or the composition of the pool, so we must
+            // retrieve the raw percentage bounds and do the full calculation. Decompress the bounds by shifting left.
+            (lowerBoundRatio, upperBoundRatio) = CircuitBreakerLib.calcBoundaryConversionRatios(
+                circuitBreakerState.decodeUint(_LOWER_BOUND_OFFSET, _BOUND_WIDTH) << _BOUND_SHIFT_BITS,
+                circuitBreakerState.decodeUint(_UPPER_BOUND_OFFSET, _BOUND_WIDTH) << _BOUND_SHIFT_BITS,
+                weightComplement
+            );
+        }
+
+        // Use the ratios retrieved (or computed) above to convert raw percentage bounds to BPT price bounds.
+        // To err in favor of tripping the breaker, round the lower bound up, and the upper bound down.
+        return (bptPrice.mulUp(lowerBoundRatio), bptPrice.mulDown(upperBoundRatio));
```

Comment:
This is the multiplication that feels like it belongs in the logic lib

Reply:
See above. I can make it call `calcBptPriceBoundaries`